File: android/spfn-ui/src/main/kotlin/xyz/superfunction/spfn/ui/NavigationHost.kt
// SPFN Mobile — the container a host app gives a pushed flow to stand in.
//
// Counterpart of Sources/SPFNUI/NavigationHost.swift: same name, same one job, same
// `HostStack` underneath.
//
// ---------------------------------------------------------------------------
// What this is for
// ---------------------------------------------------------------------------
//
// A `FlowHost(Push)` used to build a `NavDisplay` of its own and draw it over whatever the
// host app was showing. Nothing said it was wrong until a person opened one on a phone: the
// flow's first screen APPEARED rather than sliding in, and its header had no way back to the
// menu that opened it, because there was no entry under it to go back TO
// (docs/IMPLEMENTATION-PITFALLS.md P31).
//
// Decision N1 is this composable. The host app wraps whatever it draws in one
// `NavigationHost { ... }`; a `FlowHost(Push)` inside it APPENDS its routes to this back
// stack instead of building a second one, so the person gets a right-to-left push into the
// flow, the predictive back out of it, and the activity's own back at the root.
//
// ---------------------------------------------------------------------------
// A registration outlives the composition that made it, and has to
// ---------------------------------------------------------------------------
//
// `NavDisplay` composes the entries of the scene on show and no others: push a route over
// the host's root and the root's own composition — the app's menu, and every `FlowHost` in
// it — is disposed until it comes back. That is what makes a `DisposableEffect` the wrong
// place to keep this registry current, and not by a little: unregistering on dispose would
// take the flow's own entries off the stack the instant its first route covered the root,
// the stack would pop back to the root, and the two would fight for as long as anybody
// watched.
//
// So a flow registers ONCE and the store follows its stack from a coroutine of its own,
// living on this composable's scope rather than on the root's. A `Flow` publishes a
// `StateFlow`; collecting it is how the host learns about a push that happened while the
// host's root was not composed, which is every push after the first.
//
// The same sentence with the same consequence is why nothing is unregistered on dispose on
// the SwiftUI half either.
//
// ---------------------------------------------------------------------------
// What a host app puts on its own root does NOT reach a pushed flow
// ---------------------------------------------------------------------------
//
// The `root` this takes is one entry of this NavDisplay, and a pushed flow's routes are
// other entries of it. They are SIBLINGS of the root, not children of it, so a `Modifier`,
// a `CompositionLocalProvider` or a `semantics` block an app wraps its own content in is
// simply not above them.
//
// It costs a runner its selectors, silently. `testTagsAsResourceId` — the switch that turns
// a Compose test tag into the Android resource id a Maestro `id:` selector matches — is
// resolved by walking semantics PARENTS, so an app that sets it inside this host keeps it
// for its own screens and loses it for every screen of every pushed flow: the text on those
// screens still matches and every control stops being findable
// (docs/IMPLEMENTATION-PITFALLS.md P33).
//
// So anything an app means for the whole of its navigation goes OUTSIDE this composable.
// What belongs inside is what belongs to the app's own screen and not to the flows: its own
// insets, for one, because a pushed `Screen` spends the status bar inset on its own header
// where nothing above it has consumed one (docs/IMPLEMENTATION-PITFALLS.md P25).

package xyz.superfunction.spfn.ui

import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.animation.togetherWith
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.navigation3.runtime.NavEntry
import androidx.navigation3.ui.NavDisplay
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * The host app's navigation, and the back stack a pushed flow appends to.
 *
 * One per app — or one per independently navigating region — wrapped around whatever the app
 * draws:
 *
 * ```
 * NavigationHost {
 *     Box(modifier = Modifier.fillMaxSize()) {
 *         Menu();
 *         SomePushFlowHost(container);
 *     }
 * }
 * ```
 *
 * A `FlowHost` for a modal or a sheet needs nothing from this and behaves the same inside it
 * or outside it: both are presentations OVER the navigation rather than entries in it.
 *
 * `@JvmSynthetic` for the reason [FlowHost] carries it: a `@Composable` function is a rule
 * the Compose compiler enforces for Kotlin callers and for nobody else
 * (docs/IMPLEMENTATION-PITFALLS.md P15).
 */
@JvmSynthetic
@Composable
public fun NavigationHost(root: @Composable () -> Unit)
{
    val scope = rememberCoroutineScope();
    val host = remember(scope) { HostStackStore(scope) };
    val entries = host.stack.collectAsState().value.entries;

    CompositionLocalProvider(LocalNavigationHost provides host) {
        NavDisplay(
            backStack = listOf<Any>(HostRoot) + entries,
            // Only ever the top flow's, and only while there is one: on the host's own root
            // the back stack is one entry long, NavDisplay disables its own handling, and
            // the activity is what the system back closes — which is what a back on an app's
            // first screen has always meant.
            onBack = { host.back() },
            // Stated rather than left to the navigator's default, because the default is the
            // library's opinion and this one is the platform's: a push comes in from the
            // right and leaves to the right, on both platforms, and a person who does not
            // see it move does not know a screen arrived.
            transitionSpec = {
                slideInHorizontally { width -> width } togetherWith slideOutHorizontally { width -> -width / SHIFT }
            },
            popTransitionSpec = {
                slideInHorizontally { width -> -width / SHIFT } togetherWith slideOutHorizontally { width -> width }
            },
            predictivePopTransitionSpec = { _ ->
                slideInHorizontally { width -> -width / SHIFT } togetherWith slideOutHorizontally { width -> width }
            },
            entryProvider = { key ->
                if (key is HostEntry) NavEntry(key) { host.Screen(key) } else NavEntry(key) { root() }
            }
        );
    };
}

/**
 * How far the screen underneath moves while the one over it comes in.
 *
 * A fraction and not a full width: the platform's own push slides the outgoing screen a
 * short way and parallaxes it, and a screen that left at the same speed as the one arriving
 * reads as two screens passing rather than as one covering another.
 */
private const val SHIFT: Int = 4;

/**
 * The host's own root, as one key on the back stack it can never be popped past.
 *
 * A `data object` so that Navigation 3 identifies it by value the way it identifies every
 * other key, and it is deliberately not a [HostEntry]: an entry belongs to a flow and this
 * belongs to the app.
 */
private data object HostRoot

/** What one flow told the host: how to draw its routes, and what its back does. */
internal class HostRegistration(
    /**
     * Draws one of this flow's routes, wrapped in this flow's own chrome. The route arrives
     * erased, because one registry holds flows whose route types have nothing in common.
     */
    val screen: @Composable (Any) -> Unit,
    /** One back, spent through `Flow.back` like every other back in this module. */
    val back: () -> Unit
)

/**
 * The host's stack and its registry, as one object under the composition local.
 *
 * Not public: a host app builds a [NavigationHost] and a flow registers itself through
 * [FlowHost], and a third door onto this would be a second way to write a stack that has one
 * writer on purpose.
 */
internal class HostStackStore(private val scope: CoroutineScope)
{
    private val mutableStack = MutableStateFlow(HostStack());

    /** The one list, and the only thing [NavigationHost] draws from. */
    val stack: StateFlow<HostStack> = mutableStack.asStateFlow();

    private val registrations = mutableMapOf<Any, HostRegistration>();

    private val collectors = mutableMapOf<Any, Job>();

    /**
     * Says how a flow's routes are drawn and what its back does, and starts following its
     * stack.
     *
     * The registration is replaced on every call, so a host root that recomposed hands over
     * a closure that closes over what it now has. The COLLECTOR is started once: it is what
     * keeps the host's stack current while the root is not composed at all, which is the
     * whole time any of this flow's routes are on show.
     */
    fun register(owner: Any, routes: StateFlow<List<Any>>, registration: HostRegistration)
    {
        registrations[owner] = registration;
        if (collectors.containsKey(owner))
        {
            return;
        }
        collectors[owner] = scope.launch {
            routes.collect { stack -> sync(owner, stack) };
        };
    }

    /**
     * Takes one flow's stack as it now is. A closed flow syncs nothing, which is how its
     * entries leave the host's stack.
     */
    fun sync(owner: Any, routes: List<Any>)
    {
        mutableStack.value = mutableStack.value.sync(owner, routes);
    }

    /**
     * One system back, given to whoever is on top.
     *
     * Only the flow whose entry is on top can have meant it, and what it does with it is
     * `Flow.back` — the same door the header control goes through, so a gesture and a tap
     * cannot come to disagree.
     */
    fun back()
    {
        val owner = mutableStack.value.topOwner() ?: return;
        registrations[owner]?.back?.invoke();
    }

    /** Draws one entry, by asking whoever put it there. */
    @Composable
    fun Screen(entry: HostEntry)
    {
        val registration = registrations[entry.owner];
        if (registration != null)
        {
            registration.screen(entry.route);
        }
    }
}

/**
 * The host in scope, or null.
 *
 * A `FlowHost(Push)` that reads null draws its own inline stack, which is the compatibility
 * path its header describes.
 */
internal val LocalNavigationHost: ProvidableCompositionLocal<HostStackStore?> =
    staticCompositionLocalOf { null };

File: Sources/SPFNUI/NavigationHost.swift
#if canImport(SwiftUI)
// SPFN Mobile — the container a host app gives a pushed flow to stand in.
//
// Counterpart of
// android/spfn-ui/src/main/kotlin/xyz/superfunction/spfn/ui/NavigationHost.kt: same name,
// same one job, same `HostStack` underneath. Guarded whole, first line of code to last, the
// way every SwiftUI file in this module is (docs/IMPLEMENTATION-PITFALLS.md P20).
//
// ---------------------------------------------------------------------------
// What this is for
// ---------------------------------------------------------------------------
//
// A `FlowHost(.push)` used to build a `NavigationStack` of its own and draw it over
// whatever the host app was showing. Nothing said it was wrong until a person opened one on
// a phone: the flow's first screen APPEARED rather than sliding in, and its header had no
// way back to the menu that opened it, because there was no route under it to go back TO
// (docs/IMPLEMENTATION-PITFALLS.md P31). "The push root's back is the host's" was a rule
// with nothing to hand the gesture to.
//
// Decision N1 is this type. The host app wraps whatever it draws in one
// `NavigationHost { ... }`; a `FlowHost(.push)` inside it APPENDS its routes to this stack
// instead of building a second one, so the person gets the platform's own right-to-left push
// into the flow and the platform's own way back out of it.
//
// ---------------------------------------------------------------------------
// One destination, and why it is not one per flow
// ---------------------------------------------------------------------------
//
// `navigationDestination(for:)` is resolved by the TYPE of the path's elements, and this
// path carries `HostEntry` — one flow's routes and another's on one list, which is the whole
// point of the list. So there is exactly one destination, it is declared here in the host's
// own root subtree, and it asks the owner registered for that entry to draw it.
//
// The alternative — every `FlowHost` declaring `navigationDestination(for: Route.self)` at
// its own place in the tree — was considered and refused. It needs the path to carry each
// flow's concrete route type, which means `NavigationPath` and appending values whose static
// type has been erased to `AnyHashable`; whether SwiftUI then matches the destination
// registered for the route's own type is undocumented, and the failure if it does not is a
// runtime warning and a blank screen. A registration that cannot be misplaced is worth more
// than a destination declared closer to its flow: a `navigationDestination` inside a lazy
// container or outside the stack is ignored, and this one is neither by construction.
//
// What a host app wraps its own root in does NOT reach a pushed flow here either. The root
// is the stack's root and a pushed flow's routes are destinations above it, so a modifier or
// an environment value applied inside the root closure is not above them. It matters far
// less on this platform than on Compose — a SwiftUI accessibility identifier is per-view and
// survives any presentation, where Android's `testTagsAsResourceId` is inherited and does
// not (docs/IMPLEMENTATION-PITFALLS.md P33) — and the rule is the same on both: what is
// meant for the whole of an app's navigation goes outside this view.
//
// The registry is also what carries the CHROME. A `Screen` reads its way out of the
// environment, and the value it has to read is the owning flow's — not the host's, and not
// whichever flow happened to register last — so the closure the owner registers wraps its
// own content in its own chrome, read at draw time so that a stack that moved is a header
// that moved with it.

import Observation
import SwiftUI

/// The host app's navigation, and the stack a pushed flow appends to.
///
/// One per app — or one per independently navigating region — wrapped around whatever the
/// app draws:
///
/// ```swift
/// NavigationHost
/// {
///     ZStack
///     {
///         menu
///         SomePushFlowHost(container: container)
///     }
/// }
/// ```
///
/// A `FlowHost` for a modal or a sheet needs nothing from this and behaves the same inside
/// it or outside it: both are presentations OVER the navigation rather than entries in it.
@MainActor
public struct NavigationHost<Root: View>: View
{
    private let root: () -> Root

    /// The host's own stack and the flows registered against it. `@State` so it outlives
    /// every recomposition of the root, and `@Observable` so a flow appending to it is a
    /// reason for this view to redraw.
    @State private var host = HostStackStore()

    public init(@ViewBuilder root: @escaping () -> Root)
    {
        self.root = root
    }

    public var body: some View
    {
        NavigationStack(path: path)
        {
            root()
                .navigationDestination(for: HostEntry.self)
                { entry in
                    host.screen(for: entry)
                }
        }
        .environment(\.hostStack, host)
    }

    /// The host's stack, as the array `NavigationStack` calls its path.
    ///
    /// The setter is the reconciliation, and it is `FlowHost`'s own rule one level up:
    /// SwiftUI only ever shortens this array by itself — a system back, an edge swipe — so a
    /// shorter value is a pop it has already performed visually, and every flow that lost a
    /// route is told how many it lost. Any other value is ignored: pushing is `Flow.push`'s
    /// job, and accepting an arbitrary array here is precisely how a second copy of the
    /// stack starts.
    private var path: Binding<[HostEntry]>
    {
        Binding(
            get:
            {
                host.stack.entries
            },
            set:
            { newPath in
                guard newPath.count < host.stack.entries.count
                else
                {
                    return
                }
                host.shorten(to: newPath.count)
            }
        )
    }
}

/// What one flow told the host: how to draw its routes, and what its back does.
///
/// Both are closures over the flow rather than the flow itself, which is what lets one
/// registry hold flows whose route types have nothing in common.
struct HostRegistration
{
    /// Draws one of this flow's routes, wrapped in this flow's own chrome. Answers an empty
    /// view for a route that is not this flow's, which cannot happen while entries carry
    /// their owner and is still not worth a crash.
    let screen: (AnyHashable) -> AnyView

    /// One back, spent through `Flow.back(entry:)` like every other back in this module.
    let back: () -> Void
}

/// The host's stack and its registry, as one observable thing under the environment.
///
/// Not public: a host app builds a `NavigationHost` and a flow registers itself through
/// `FlowHost`, and a third door onto this would be a second way to write a stack that has
/// one writer on purpose.
@MainActor
@Observable
final class HostStackStore
{
    /// The one list, and the only thing `NavigationHost` draws from.
    private(set) var stack = HostStack()

    private var registrations: [ObjectIdentifier: HostRegistration] = [:]

    /// Says how a flow's routes are drawn and what its back does. Re-registering is how a
    /// flow keeps its chrome current, so this replaces rather than refuses.
    func register(owner: ObjectIdentifier, registration: HostRegistration)
    {
        registrations[owner] = registration
    }

    /// Takes one flow's stack as it now is. A closed flow syncs nothing, which is how its
    /// entries leave the host's stack.
    func sync(owner: ObjectIdentifier, routes: [AnyHashable])
    {
        stack = stack.sync(owner: owner, routes: routes)
    }

    /// Tells every flow that lost routes to a platform pop how many it lost.
    ///
    /// The flows are what shorten the stack: each `back` moves that flow's own state, the
    /// flow publishes it, and the entries follow. Nothing here writes `stack` directly, for
    /// the reason `FlowHost` never does either.
    func shorten(to count: Int)
    {
        for (owner, dropped) in stack.shortened(to: count)
        {
            guard let registration = registrations[owner]
            else
            {
                continue
            }
            for _ in 0 ..< dropped
            {
                registration.back()
            }
        }
    }

    /// Draws one entry, by asking whoever put it there.
    func screen(for entry: HostEntry) -> AnyView
    {
        registrations[entry.owner]?.screen(entry.route) ?? AnyView(EmptyView())
    }
}

private struct HostStackKey: EnvironmentKey
{
    /// No host. A `FlowHost(.push)` that reads this draws its own inline stack, which is
    /// the compatibility path its header describes.
    static let defaultValue: HostStackStore? = nil
}

extension EnvironmentValues
{
    var hostStack: HostStackStore?
    {
        get { self[HostStackKey.self] }
        set { self[HostStackKey.self] = newValue }
    }
}
#endif
